test/e2e/e2e_test.go

```diff
@@ -301,4 +303,60 @@ var _ = ginkgo.Describe("leaderWorkerSet e2e tests", func() {
 			return numberOfPodsInCommon, nil
 		}, timeout, interval).Should(gomega.Equal(0))
 	})
+	ginkgo.It("Not updated worker StatefulSet restarted during rolling update will be restored with old worker spec", func() {
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(2).Size(2).Obj()
+		testing.MustCreateLws(ctx, k8sClient, lws)
+		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
+
+		testing.ExpectValidLeaderStatefulSet(ctx, k8sClient, lws, 2)
+		testing.UpdateWorkerTemplate(ctx, k8sClient, lws)
+		gomega.Expect(k8sClient.Delete(ctx, &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: lws.Namespace, Name: lws.Name + "-0-1"}})).To(gomega.Succeed())
+		gomega.Eventually(func() error {
+			var sts appsv1.StatefulSet
+			if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name + "-0", Namespace: lws.Namespace}, &sts); err != nil {
+				return err
+			}
+			// use the original lws object instead of newest, since we are comparing with the old worker template
+			podTemplateSpec := *lws.Spec.LeaderWorkerTemplate.WorkerTemplate.DeepCopy()
+			if sts.Spec.Template.Spec.Containers[0].Name != podTemplateSpec.Spec.Containers[0].Name {
+				return fmt.Errorf("StatefulSet did not have the expected container name")
+			}
+			return nil
+		}, timeout, interval).Should(gomega.Succeed())
```

Comment:
This test will be flaky, how do you guarantee that the update isn't faster than your check here?

Reply:
There is no guarantee, that is just a risk that is taken when checking the state mid update. Similar issue in other e2e tests. All I can do is have as many replicas as possible so that the update takes long to reach the "-0" leader sts.

Reply:
We can't have tests that don't offer guarantees, those will flake, which means when they fail, we don't know if they fail because of a bug or the race condition. The test you are trying to do here should be done via an integration test  where we control update progress.

Reply:
Right, but other update tests also don't offer guarantees. For example this one https://github.com/kubernetes-sigs/lws/blob/main/test/e2e/e2e_test.go#L181, what guarantees that the update isn't done before this check is performed? I guess I'm just confused as to why we can do it there but not here, both are checking the state mid update.